File: leetcode/128-longest-consecutive-sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/ (top-interview-questions)

from typing import List

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numSet = set(nums)
        maxLength = 0
        
        for num in numSet:
            if (num - 1) not in numSet:
                length = 1
                while (num + length) in numSet: length += 1
                maxLength = max(maxLength, length)
        return maxLength
    

    def longest_Consecutive(self, nums: List[int]) -> int:
        if len(nums) == 0: return 0

        numSet = set(nums)
        longestSequence = 1

        for num in nums:
            if num not in numSet:
                continue

            l = num - 1
            r = num + 1
            while l in numSet or r in numSet:
                if l in numSet:
                    numSet.remove(l)
                    l -= 1
                if r in numSet:
                    numSet.remove(r)
                    r += 1
            
            longestSequence = max(longestSequence, (r - l - 1))
        
        return longestSequence

File: leetcode/128-longest-consecutive-sequence.js
// https://leetcode.com/problems/longest-consecutive-sequence/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function(nums) {
  if (nums.length === 0) return 0;
  
  let longestSequence = 1;
  const hashSet = new Set(nums);

  for (const num of nums) {
    if (!hashSet.has(num)) continue;
    
    hashSet.delete(num);
    let l = num - 1;
    let r = num + 1;
    while (hashSet.has(l) || hashSet.has(r)) {
      if (hashSet.has(l)) {
        hashSet.delete(l);
        l -= 1;
      }
      if (hashSet.has(r)) {
        hashSet.delete(r);
        r += 1;
      }
    }

    longestSequence = Math.max(longestSequence, (r - l - 1));
  }

  return longestSequence;

};


var longestConsecutive = function(nums) {
  
  let longestSequence = 0;

  const hashSet = new Set(nums);

  for (const num of nums) {
    if (hashSet.has(num - 1)) {
      // it's not a start of consecutive sequence
      continue;
    }

    // num is a start of a consecutive sequence, let's compute her length
    let length = 1;
    while (hashSet.has(num + length)) {
      length += 1;
    }
    longestSequence = Math.max(longestSequence, length);
  }

  return longestSequence;
}
